discovery-infra/test_infra/controllers/node_controllers/node_controller.py

```diff
@@ -1,73 +1,114 @@
-from typing import Dict, List, Any
-from test_infra.controllers.node_controllers import node
+from typing import Dict, List, Any, Tuple
+from test_infra.controllers.node_controllers.node import Node
+from abc import ABC, abstractmethod
 
 
-class NodeController:
-    def list_nodes(self) -> Dict[str, node.Node]:
-        raise NotImplementedError
+class NodeController(ABC):
```

Comment:
ABC is better than no ABC, and this is how you work with ABC. ABC makes sure you don't even instantiate a class that doesn't have all its (and its parents) abstract methods implemented. You clearly mark methods as abstract with `@abstractmethod`. Since it won't even let you instantiate the class, it will catch missing implementation errors much more early. If you only raise when the method is called, you only catch this bug during runtime, and if that method is not called on your happy path, that bug can hide for a long time before you even figure out that it's missing

Reply:
FYI @tsorya @nshidlin  @lalon4